textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/keywordmatch/KeywordMatcher.java

```diff
@@ -0,0 +1,168 @@
+                spanSchema = Utils.createSpanSchema(schema);
+            }
+
+            for(int attributeIndex = 0; attributeIndex < attributeList.size(); attributeIndex++){
+                IField field = sourceTuple.getField(attributeList.get(attributeIndex).getFieldName());
+                String fieldValue = (String) (field).getValue();
+                if(field instanceof StringField){
+                    //Keyword should match fieldValue entirely
+
+                    if(fieldValue.equals(queryValue.toLowerCase())){
+                        spanIndexValue = 0;
+                        positionIndex = queryValue.length();
+                        fieldName = attributeList.get(attributeIndex).getFieldName();
+                        addSpanToSpanList(fieldName, spanIndexValue, positionIndex, queryValue, fieldValue);
+                        foundFlag = true;
+                    }
+                }
+                else if(field instanceof TextField) {
+                    //Each element of Array of keywords is matched in tokenized TextField Value
+                    for (int iter = 0; iter < queryValueArray.size(); iter++) {
+                        positionIndex = 0;
+                        String query = queryValueArray.get(iter);
+                        Pattern p = patternList.get(iter);
+                        matcher = p.matcher(fieldValue.toLowerCase());
+                        while (matcher.find(positionIndex) != false) {
+                            spanIndexValue = matcher.start();
+                            positionIndex = spanIndexValue + queryValueArray.get(iter).length();
+                            documentValue = fieldValue.substring(spanIndexValue, positionIndex);
+                            fieldName = attributeList.get(attributeIndex).getFieldName();
+                            addSpanToSpanList(fieldName, spanIndexValue, positionIndex, query, documentValue);
+                            foundFlag = true;
+
+                        }
+                    }
+                }
+                positionIndex = 0;
+            }
+
+            //If all the 'attributes to be searched' have been processed return the result tuple with span info
+            if (foundFlag){
+                foundFlag = false;
+                positionIndex = 0;
+                return Utils.getSpanTuple(fieldList, spanList, spanSchema);
+            }
+            //Search next document if the required predicate did not match previous document
+            else if(sourceTuple != null) {
+                positionIndex = 0;
+                spanList.clear();
+
+                return getNextTuple();
+
+            }
+
+            return null;
+
+        } catch (Exception e) {
+            e.printStackTrace();
+            throw new DataFlowException(e.getMessage(), e);
+        }
+
+    }
+
+    private void addSpanToSpanList(String fieldName, int start, int end, String key, String value) {
```

Comment:
It is there in `DictionaryMatcher`. But it is a very small function, so I didn't expose it in `Utils` class